agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -2666,6 +2669,54 @@ bool backhaul_manager::handle_1905_combined_infrastructure_metrics(
     return send_cmdu_to_bus(cmdu_tx, src_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::handle_1905_topology_discovery(const std::string &src_mac,
+                                                      ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto tlvAlMac = cmdu_rx.getClass<ieee1905_1::tlvAlMacAddressType>();
+    if (!tlvAlMac) {
+        LOG(ERROR) << "addClass tlvAlMacAddressType failed";
+        return false;
+    }
+
+    // Filter out the messages we have sent.
+    if (tlvAlMac->mac() == network_utils::mac_from_string(bridge_info.mac)) {
+        return true;
+    }
+
+    auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received TOPOLOGY_DISCOVERY_MESSAGE from AL MAC=" << tlvAlMac->mac()
+              << ", mid=" << std::hex << int(mid);
+
+    auto new_device = m_1905_neighbor_devices.find(tlvAlMac->mac()) == m_1905_neighbor_devices.end()
+                          ? true
+                          : false;
+
+    // Add/Update the device on our list.
+    m_1905_neighbor_devices[tlvAlMac->mac()].last_seen = std::chrono::steady_clock::now();
+
+    auto tlvMac = cmdu_rx.getClass<ieee1905_1::tlvMacAddress>();
+    if (!tlvMac) {
+        LOG(ERROR) << "Failed to getClass tlvMacAddress";
+        return false;
+    }
+    m_1905_neighbor_devices[tlvAlMac->mac()].local_iface_mac = tlvMac->mac();
```

Comment:
tlvMac->mac() is the peer iface mac, not the local iface mac.

Unfortunately, the local iface information is discarded by transport_socket_thread::handle_cmdu_message_bus(). It is available in `cmdu_rx_msg->metadata()->if_index` but that doesn't get forwarded into handle_cmdu(). So some more refactoring will be needed to be able to properly populate the neighbour list in the topology response.

Reply:
How it can be the peer iface mac? (by peer you mean neighbor al_mac right?)
Because the other tlv is al_mac_tlv.

Reply:
peer iface mac is the MAC address of the interface on which the neighbour sends out its topology discovery messages. I thought it was part of the 1905.1 neighbour TLV, but apparently I was mistaken - it's only included in the link metrics.

If there is a single link between two neighbours, you have four MAC addresses involved: the AL-MAC address of this device, the AL-MAC address of the neighbour device, the MAC address of the interface on which the topology discovery is received by this device, and the MAC address of the interface on which the topology discovery message is sent by the neighbour device. The AL-MAC address and the sending interface MAC address of the neighbour device come from the topology discover message - AL MAC address type resp. MAC address type TLV. The local interface on which the topology discovery message was received has to be recorded by the receiving AL Entity, but, as mentioned, currently we don't.

Reply:
Currently, we fill the interface mac on the discovery message with our bridge mac.
Second, I don't understand how could we know from each interface the message will be sent with. I mean, we send something to the bus. We don't get information about the routes it used to send the message. Also, the Discovery message uses multicast. It could, in theory, use all available interfaces, but we could only fill one interface.
It is not so clear to me.